components/webdriver_server/actions.rs

```diff
@@ -317,5 +358,148 @@ impl Handler {
                 .send(ConstellationMsg::WebDriverCommand(cmd_msg))
                 .unwrap();
         }
+
+        Ok(())
+    }
+
+    // https://w3c.github.io/webdriver/#dfn-dispatch-a-pointermove-action
+    fn dispatch_pointermove_action(
+        &mut self,
+        source_id: &str,
+        action: &PointerMoveAction,
+        tick_duration: u64,
+    ) -> Result<(), ErrorStatus> {
+        let tick_start = Instant::now();
+
+        let (start_x, start_y) = match self
+            .session
+            .as_ref()
+            .unwrap()
+            .input_state_table
+            .get(source_id)
+            .unwrap()
+        {
+            InputSourceState::Null => unreachable!(),
+            InputSourceState::Key(_) => unreachable!(),
+            InputSourceState::Pointer(pointer_input_state) => {
+                (pointer_input_state.x as i64, pointer_input_state.y as i64)
+            },
+        };
+
+        let x_offset = action.x.unwrap_or(0);
+        let y_offset = action.y.unwrap_or(0);
+
+        let (x, y) = match action.origin {
+            PointerOrigin::Viewport => (x_offset, y_offset),
+            PointerOrigin::Pointer => (start_x + x_offset, start_y + y_offset),
+            PointerOrigin::Element(ref x) => {
+                let (sender, receiver) = ipc::channel().unwrap();
+                self.top_level_script_command(WebDriverScriptCommand::GetElementInViewCenterPoint(
+                    x.to_string(),
+                    sender,
+                ))
+                .unwrap();
+
+                match receiver.recv().unwrap() {
+                    Ok(point) => match point {
+                        Some((x_element, y_element)) => (x_element, y_element),
+                        None => return Err(ErrorStatus::UnknownError),
+                    },
+                    Err(_) => return Err(ErrorStatus::UnknownError),
+                }
+            },
+        };
+
+        let (sender, receiver) = ipc::channel().unwrap();
+        let cmd_msg = WebDriverCommandMsg::GetWindowSize(
+            self.session.as_ref().unwrap().top_level_browsing_context_id,
+            sender,
+        );
+        self.constellation_chan
+            .send(ConstellationMsg::WebDriverCommand(cmd_msg))
+            .unwrap();
+
+        let viewport = receiver.recv().unwrap().initial_viewport;
+        if x < 0 || x as f32 > viewport.width || y < 0 || y as f32 > viewport.height {
+            return Err(ErrorStatus::MoveTargetOutOfBounds);
+        }
+
+        let duration = match action.duration {
+            Some(duration) => duration,
+            None => tick_duration,
+        };
+
+        thread::sleep(Duration::from_millis(17));
```

Comment:
What is the purpose of this sleep?

Reply:
This is step 9.1 from https://w3c.github.io/webdriver/#dfn-process-a-pointer-move-action. The actual value of `17` milliseconds is taken from https://searchfox.org/mozilla-central/source/testing/marionette/action.js#1377. I'll add a comment describing its meaning.